src/vfs2.c

```diff
@@ -308,93 +339,148 @@ static bool wal_index_basic_hdr_advanced(struct vfs2_wal_index_basic_hdr new,
 	       && new.mxFrame > old.mxFrame;
 }
 
+/* TODO this is very weak, should check that the hash tables have been constructed */
+static bool wal_index_recovered(struct entry *e)
+{
+	PRE(e->shm_regions_len > 0);
+	char *p = e->shm_regions[0];
+	for (size_t i = sizeof(struct wal_index_full_hdr); i < VFS2_WAL_INDEX_REGION_SIZE; i++) {
+		if (p[i] != 0) {
+			return true;
+		}
+	}
+	return false;
+}
+
+static bool is_valid_page_size(unsigned long n)
+{
+	return n >= 1 << 9 && n <= 1 << 16 && (n & (n - 1)) == 0;
+}
+
+static bool is_open(struct entry *e)
+{
+	return e->main_db_name != NULL
+		&& e->wal_moving_name != NULL
+		&& e->wal_cur_fixed_name != NULL
+		&& e->wal_cur != NULL
+		&& e->wal_prev_fixed_name != NULL
+		&& e->wal_prev != NULL
+		&& (e->refcount_main_db > 0 || e->refcount_wal > 0)
+		&& e->shm_regions != NULL
+		&& e->shm_regions_len > 0
+		&& e->shm_regions[0] != NULL
+		&& e->common != NULL;
+}
+
+static bool basic_hdr_valid(struct wal_index_basic_hdr bhdr)
```

Comment:
Happy to do this, but doesn't this conversion happen anyway when lowering to the SysV ABI? Is it just about avoiding stack-to-stack copies?

Reply:
It's easier to reason about a function properties with `struct wal_index_basic_hdr*` because it's not needed to think about the future of `struct wal_index_basic_hdr`.